acquire/dynamic/windows/handles.py

```diff
@@ -0,0 +1,255 @@
+
+advapi32 = ctypes.windll.advapi32
+advapi32.OpenProcessToken.argtypes = [HANDLE, DWORD, PHANDLE]
+
+kernel32 = ctypes.windll.kernel32
+kernel32.OpenProcess.restype = HANDLE
+kernel32.DuplicateHandle.argtypes = (HANDLE, HANDLE, HANDLE, ctypes.POINTER(HANDLE), DWORD, BOOL, DWORD)
+
+
+class DuplicateFilter(Filter):
+    def __init__(self) -> None:
+        super().__init__()
+        self.msgs = set()
+
+    def filter(self, record: LogRecord) -> bool:
+        seen = (msg := record.getMessage()) not in self.msgs
+        self.msgs.add(msg)
+        return seen
+
+
+def get_handle_type_info(handle: SYSTEM_HANDLE_TABLE_ENTRY_INFO_EX) -> Optional[str]:
+    """Return type of handle.
+
+    Parameters:
+        handle: handle for which to return the type information.
+
+    Raises:
+        RuntimeError: Raised when the result of the object query is unknown (other than SUCCESS, LENGTH MISMATCH or
+        INVALID).
+    """
+    public_object_type_information = PUBLIC_OBJECT_TYPE_INFORMATION()
+    size = DWORD(ctypes.sizeof(public_object_type_information))
+    while True:
+        result = ntdll.NtQueryObject(
+            handle,
+            OBJECT_INFORMATION_CLASS.ObjectTypeInformation,
+            ctypes.byref(public_object_type_information),
+            size,
+            None,
+        )
+
+        if result == NtStatusCode.STATUS_SUCCESS:
+            return str(public_object_type_information.Name)
+        elif result == NtStatusCode.STATUS_INFO_LENGTH_MISMATCH:
+            size = DWORD(size.value * 4)
+            ctypes.resize(public_object_type_information, size.value)
+        elif result == NtStatusCode.STATUS_INVALID_HANDLE:
+            return None
+        else:
+            raise RuntimeError(hex(result))
+
+
+def _get_file_name_thread(h_file: HANDLE, q: Queue):
+    iob = IO_STATUS_BLOCK()
+    file_name_information = ctypes.create_string_buffer(0x1000)
+
+    result = ntdll.NtQueryInformationFile(
+        h_file,
+        ctypes.byref(iob),
+        file_name_information,
+        len(file_name_information),
+        FILE_INFORMATION_CLASS.FileNameInformation,
+    )
+
+    file_name = None
+    if result == NtStatusCode.STATUS_SUCCESS:
+        file_name_length = struct.unpack("<I", file_name_information[:4])[0]
+        file_name = file_name_information[4 : 4 + file_name_length].decode("utf-16-le")
+
+    q.put(file_name)
+
+
+def get_handle_name(pid: int, handle: SYSTEM_HANDLE_TABLE_ENTRY_INFO_EX) -> Optional[str]:
+    """Return handle name."""
+
+    remote = pid != kernel32.GetCurrentProcessId()
+
+    if remote:
+        try:
+            h_remote = open_process(pid)
+        except OpenProcessError as e:
+            log.error(e)
+            return None
+        try:
+            handle = duplicate_handle(h_remote, handle)
+        except RuntimeError:
+            close_handle(h_remote)
+            return None
+
+    q = Queue()
+    thread = threading.Thread(target=_get_file_name_thread, args=(handle, q))
```

Comment:
Why is this done in a thread?

Reply:
Added a comment to explain why a thread is used here ;)